bluesky_widgets/qt/search_results.py

```diff
@@ -94,55 +53,55 @@ def __init__(self, model, *args, **kwargs):
         self._current_num_rows = 0
         self._catalog_length = len(self.model.catalog)
 
-        # State related to asynchronously fetching data
+        # Cache for loaded data
         self._data = {}
-        self._request_queue = queue.Queue()
-
-        # This is a QThread that will run the blocking operations required to
-        # fetch data.
-        data_loader = DataLoader(
-            self._request_queue,
-            self.model.get_data,
-            self._data,
-            self.dataChanged,
-            parent=None,
-        )
-
-        def stop_data_loader():
-            # This must not contain references to self, or finalize will never
-            # be called.
-            data_loader.requestInterruption()
-            logger.debug("Initiated graceful shutdown of DataLoader")
-            # Wait for the polling loop in DataLoader to process the interruption
-            # request.
-            data_loader.wait()
-            data_loader.deleteLater()
-
-        # When this Python object is destroyed by Qt, gracefully stop the
-        # DataLoader QThread before Qt destroys it.
-        weakref.finalize(self, stop_data_loader)
+        # Queue of indexes of data to be loaded
+        self._work_queue = collections.deque()
+        # Set of active workers
+        self._active_workers = set()
 
-        # The DataLoader thread is not started here. It will be started if/when
-        # we need it for the first time.
-        self._data_loader = data_loader
+        # Start a timer that will periodically load any data queued up to be loaded.
+        LATENCY = 100  # ms
+        self._data_loading_timer = QTimer(self)
+        self._data_loading_timer.timeout.connect(self._process_work_queue)
+        self._data_loading_timer.start(LATENCY)
```

Comment:
This is great. Entirely optional alternative strategy, is to create a single shot Timer and in self._process_work_queue (restart the timer). That way you don't run into an issue where Timer keeps firing _process_work_queue one after another in case the _process_work_queue ever takes greater than LATENCY

Reply:
Good point. @tacaswell pointed me in this direction in some much earlier draft of this, but it fell out of my brain. Thanks for the reminder.

Reply:
Addressed in 6ea3841